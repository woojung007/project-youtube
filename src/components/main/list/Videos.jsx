import { useQuery } from '@tanstack/react-query';
import React from 'react';
import { fetchVideosByKeyword } from '../../layout/header/SearchHeader.queries';
import styles from './Videos.module.css';
import VideoCard from './item/VideoCard';

export default function Videos() {
    const { data: videos, isLoading, error } = useQuery(['videos', ''], fetchVideosByKeyword);
    // fetchVideos

    if (isLoading) return <div>loading ...</div>;

    if (error) return <div> {error}</div>;

    return (
        <div className={styles.container}>
            {videos?.map((video, idx) => (
                <VideoCard key={idx} video={video} />
            ))}
        </div>
    );
}
