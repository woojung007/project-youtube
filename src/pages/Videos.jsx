import { useQuery } from '@tanstack/react-query';
import FakeYoutube from 'api/fakeYoutube';
import Youtube from 'api/youtube';
import VideoCard from 'components/VideoCard';
import React from 'react';
import { useParams } from 'react-router-dom';

export default function Videos() {
    const { keyword } = useParams();
    const {
        isLoading,
        error,
        data: videos,
    } = useQuery(['videos', keyword], () => {
        const youtube = new Youtube();
        return youtube.search(keyword);
    });

    return (
        <>
            <div>videos {keyword ? `${keyword}` : 'ðŸ”¥'}</div>

            {isLoading && <p>Loading...</p>}
            {error && <p>Something is wrong ðŸ˜…</p>}
            {videos && (
                <ul>
                    {videos.map((video) => (
                        <VideoCard key={video.id} video={video} />
                    ))}
                </ul>
            )}
        </>
    );
}
