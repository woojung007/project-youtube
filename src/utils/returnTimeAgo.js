import moment from '../../node_modules/moment/moment';

export const returnTimeAgo = (date) => {
    const now = new Date();

    return moment(date).from(moment(now));
};
