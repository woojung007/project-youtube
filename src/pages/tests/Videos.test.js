import { fakeVideo, fakeVideos } from 'tests/videos';
import { render, waitFor, screen, waitForElementToBeRemoved } from '@testing-library/react';
import { withAllContexts, withRouter } from 'tests/utils';
import Videos from 'pages/Videos';
import { Route } from 'react-router-dom';
import '@testing-library/jest-dom';

describe('Videos', () => {
    const fakeYoutube = {
        search: jest.fn(),
    };

    beforeEach(() => {
        fakeYoutube.search.mockImplementation((keyword) => {
            return keyword ? [fakeVideo] : fakeVideos;
        });
    });

    afterEach(() => {
        fakeYoutube.search.mockReset();
    });

    it('renders correctly - Snapshot', async () => {
        const { asFragment } = renderWithPath('/');

        await waitForElementToBeRemoved(screen.queryByText('Loading...'));
        expect(asFragment()).toMatchSnapshot();
    });

    it('renders all videos when keyword is not specified', async () => {
        renderWithPath('/');

        expect(fakeYoutube.search).toHaveBeenCalledWith(undefined);
        await waitFor(() => expect(screen.getAllByRole('listitem')).toHaveLength(fakeVideos.length));
    });

    it('when keyword is specified, renders search keyword', async () => {
        const searchKeyword = 'fake-keyword';
        renderWithPath(`/${searchKeyword}`);

        expect(fakeYoutube.search).toHaveBeenCalledWith(searchKeyword);
        await waitFor(() => expect(screen.getAllByRole('listitem')).toHaveLength(1));
    });

    it('renders loading state when items are being fetched', async () => {
        renderWithPath('/');

        expect(screen.getByText(/Loading.../i)).toBeInTheDocument();
    });

    it('renders error state when fetching items fails', async () => {
        fakeYoutube.search.mockImplementation(async () => {
            throw new Error('error');
        });

        renderWithPath('/');

        await waitFor(() => {
            expect(screen.getByText(/Something is wrong/i)).toBeInTheDocument();
        });
    });

    function renderWithPath(path) {
        return render(
            withAllContexts(
                withRouter(
                    <>
                        <Route path='/' element={<Videos />} />
                        <Route path='/:keyword' element={<Videos />} />
                    </>,
                    path,
                ),
                fakeYoutube,
            ),
        );
    }
});
