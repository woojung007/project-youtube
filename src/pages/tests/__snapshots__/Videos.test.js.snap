// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Videos renders correctly - Snapshot 1`] = `
<DocumentFragment>
  <ul
    class="grid grid-cols-1 sm:grid-cols-2 lg:grid-cols-3 xl:grid-cols-4 2xl:grid-cols-5 gap-2 gap-y-4"
  >
    <li
      class=""
    >
      <img
        alt="title"
        class="w-full"
        src="http://image/"
      />
      <div>
        <p
          class="my-2 font-semibold line-clamp-2"
        >
          title
        </p>
        <p
          class="text-sm opacity-80"
        >
          channelTitle
        </p>
        <p
          class="text-sm opacity-80"
        >
          방금
        </p>
      </div>
    </li>
    <li
      class=""
    >
      <img
        alt="title2"
        class="w-full"
        src="http://image2/"
      />
      <div>
        <p
          class="my-2 font-semibold line-clamp-2"
        >
          title2
        </p>
        <p
          class="text-sm opacity-80"
        >
          channelTitle2
        </p>
        <p
          class="text-sm opacity-80"
        >
          방금
        </p>
      </div>
    </li>
  </ul>
</DocumentFragment>
`;
